### DISCUSSION TITLE
How to generate files for the application in real time ?

### DISCUSSION THREAD

**Discussion by jonas-martinez** [lenra-io/lenra_cli#108] [General]
Let's say that I create an image editing application on Lenra. The problem is when I want to upload an image from the application, we cannot store the image on the application because the template doesn't have write access inside the `resources/` folder. Same for editing.

@lenra-io/lenra

**Comment by jonas-martinez** [comment]
Do we want to give write access to the `resources` folder or let the developer use an external service to store his files ?

**Reply by shiipou** [reply]
Granting write access to the directory `resources` will make thing a little more complicated to save theses files across the cluster.
An app with a `volume` linked to it will lock this volume to the current `node` of the pod in the cluster.
That mean an app with multiple pods for better performance can run only on the same node.
To allow it, we need to add other (external) storage service. (like `minio`/`openio`, `glusterfs` or `ceph`)

**Reply by taorepoara** [reply]
I don't think so. This kind of resources have to be handled differently.

**Comment by jonas-martinez** [comment]
If we store the files on Lenra, how do we handle an application with an infinite loop creating tons and tons of files ?

**Reply by taorepoara** [reply]
I think we can set a store size limit

**Comment by Nesqwik** [comment]
Files (including images) will probably necessitate an object storage to store them. 

We will provide an API in the future for this use case with limitation and biling for more storage if needed.
Currently any developper can use a similar service without us since the pods are not closed from outside.

**Reply by shiipou** [reply]
I think you're right, object storage like `minio` (An opensource service compatible with S3 object storage) can be the best way to handle it.
But that need a little more configuration for the developer. (Most language has package to interact with `S3` services)

**Reply by Nesqwik** [reply]
OVH can provide an object storage too : https://www.ovhcloud.com/fr/public-cloud/object-storage/

**Reply by taorepoara** [reply]
I think the OVH solution is the best one

**Reply by shiipou** [reply]
> OVH can provide an object storage too : https://www.ovhcloud.com/fr/public-cloud/object-storage/

@Nesqwik You're right, I forgot it.

**Reply by Nesqwik** [reply]
Please note that the goal of the Feature request is to talk about the feature itself (how to use it, what is needed), not the technical aspect (how to develop it) ! :)

**Comment by taorepoara** [comment]
I think that if a user upload a file in a Lenra app he can have the possibility to download it back throught the same app and if it's an image (or font or sound, etc when we handle it) the file should be usable by the application just like a file in the `resources` folder

**Reply by Nesqwik** [reply]
If we simplify the Openfaas Template to give the developer the responsibility to route the entire POST / request, he basically can choose what the `resource query` should respond. The dev can then return the file in the `resource directory` as we do now or virtually return anything he wants.
With this system the developer can for example : 
- Use an external file storage system (FTP, webdav, HTTP...) and stream back the file
- Send back any file from the resource directory
- stream back the file stored in the future "Lenra object storage" 
- Send back any binary data that can be processed in memory as a file

Of course we can (and should !) provide the code to return static file or "Lenra object storage" files as example.

This is more complexe than just give a simple Object storage API that handle most of the complexity.

As always, it's "Permissive" vs "Easy to use"...

**Reply by taorepoara** [reply]
It's a good point of view !
With this system we'll have to let the developer define cache for the resource to avoid app overload for nothing.

**Reply by taorepoara** [reply]
With the language agnostic generation every thing is managed by the app, so we maybe can remove the load of the manifest at the app start ?

**Reply by Nesqwik** [reply]
There is some uses for a "manifest" load at start, especially routes and "base" components.

This is an example of a manifest using the "external" view system and a possible "page" system for the Lenra view.
```javascript
module.exports = () => {
  return {
    routes: {
      "/counters": {
        name: "countersJson",
        query: {
          "$find": {
            "_datastore": "Counter"
          }
        },
        view: 'json'
      },
      "/counters/:id": {
        name: "counterJson",
        query: {
          "$find": {
            "_datastore": "Counter",
            "_id": "@id"
          }
        },
        view: 'json'
      },
      "/lenra": {
        name: "main",
        view: "lenra"
      },
    },
  }
}
```

This allow us to prepare a query for the first widget, tell the system which view you want to use etc..
This way you can mix the view depending of the usage. You can for example build a UI using React and "json" view for the users and a fast and easy view using the "lenra" view for the admin panel of your app.

Note that we don't have to keep the name "manifest" but change it to "routes". 
If the "routes" are called for the session every time (and not the environment), it allow the dev to generate different routes based on the user data (for example, not generating the admin routes for a non admin user).
